### DISCUSSION TITLE
#if HAS_ or #ifdef HAS_ standardisation

### DISCUSSION THREAD

**Discussion by S5NC** [meshtastic/firmware#3632] [General]
There are 32 uses of `#if HAS_[...]` and 28 of `#ifdef HAS_[...]`, we should standardise on one.
We would need to make sure `#elif`, `#ifndef`, etc. are also made uniform and `#else` behaviours are consistent before and after the change.

I am in favour of using `#if HAS_`, and having a file, as we currently do, which has the default 1 or 0 values. This alleviates the need to make `HAS_[...]` keywords have a polarity of the default option, and we can make all `HAS_[...]` names positive, so 1 clearly enables, and 0 disables that feature.

**Comment by jp-bennett** [comment]
+1 as I've recently found a minor bug resulting from an accidental ifdef HAS_.

**Comment by mrx23dot** [comment]
Definitely **#if**, that way compiler complains if headers are messed up.
Common mistake rookies do.

**Comment by oseiler2** [comment]
Small contribution towards this: https://github.com/meshtastic/firmware/pull/3733

**Comment by S5NC** [comment]
https://github.com/search?q=repo%3Ameshtastic%2Ffirmware+%22%23ifdef+HAS_%22&type=code
https://github.com/search?q=repo%3Ameshtastic%2Ffirmware+%22defined%28HAS_%22&type=code
https://github.com/search?q=repo%3Ameshtastic%2Ffirmware+%22%23ifndef+HAS_%22&type=code

I think the only place for definition checks should be in configuration.h like
```cpp
#ifndef HAS_WIFI
#define HAS_WIFI 0
#endif
```

- Have global **default** behaviours, which set each `HAS_[...]` to `0` or `1`, probably disabling all, in `configuration.h`
- Perhaps have default behaviours for each architecture or board?
- Each `variant.h` file can set the `HAS_[...]` they want to change
- Use `#ifndef` working backwards, first including `variant.h`, then `configuration.h`, because having `#undef`s in `variant.h` is a complication.
